state/optimistic_syncing/pipeline.go

```diff
@@ -0,0 +1,424 @@
+// from the latest persisted result.
+func (p *Pipeline) broadcastStateUpdate() {
+	if p.childrenUpdateChan == nil {
+		return
+	}
+
+	p.mu.RLock()
+	// Only descend from latest if this pipeline descends AND is not canceled
+	descendsForChildren := p.descendsFromSealed && p.state != StateCanceled
+	update := StateUpdate{
+		DescendsFromLastPersistedSealed: descendsForChildren,
+		ParentState:                     p.state,
+	}
+	p.mu.RUnlock()
+
+	p.childrenUpdateChan <- update
+}
+
+// processCurrentState handles the current state and transitions to the next state if possible.
+// It returns false when a terminal state is reached (StateComplete or StateCanceled), true otherwise.
+func (p *Pipeline) processCurrentState(ctx context.Context) bool {
+	p.mu.RLock()
+	currentState := p.state
+	p.mu.RUnlock()
+
+	switch currentState {
+	case StateReady:
+		return p.processReady()
+	case StateDownloading:
+		return p.processDownloading(ctx)
+	case StateIndexing:
+		return p.processIndexing(ctx)
+	case StateWaitingPersist:
+		return p.processWaitingPersist()
+	case StatePersisting:
+		return p.processPersisting(ctx)
+	case StateComplete, StateCanceled:
+		// Terminal states
+		return false
+	default:
+		p.logger.Error().Str("state", currentState.String()).Msg("invalid state")
+		return false
+	}
+}
+
+// transitionTo transitions the pipeline to the given state and broadcasts
+// the state change to children pipelines.
+func (p *Pipeline) transitionTo(newState State) {
+	p.mu.Lock()
+	oldState := p.state
+	p.state = newState
+	p.mu.Unlock()
+
+	p.logger.Info().
+		Str("oldState", oldState.String()).
+		Str("newState", newState.String()).
+		Msg("pipeline state transition")
+
+	// Broadcast state update to children
+	go p.broadcastStateUpdate()
+
+	// Trigger state check in case we can immediately transition again
+	if newState != StateComplete && newState != StateCanceled {
+		p.stateNotifier.Notify()
+	}
+}
+
+// processReady handles the Ready state and transitions to StateDownloading if possible.
+// Returns true to continue processing, false if a terminal state was reached.
+func (p *Pipeline) processReady() bool {
+	if p.canStartDownloading() {
+		p.transitionTo(StateDownloading)
+		return true
+	}
+	return true
+}
+
+// processDownloading handles the Downloading state.
+// It executes the download function and transitions to StateIndexing if possible.
+// Returns true to continue processing, false if a terminal state was reached.
+//
+// No error returns during normal operations.
+func (p *Pipeline) processDownloading(ctx context.Context) bool {
+	p.logger.Info().Msg("starting download step")
+
+	if err := p.downloadFunc(ctx); err != nil {
+		p.logger.Error().Err(err).Msg("download step failed")
+		p.transitionTo(StateCanceled)
+		return false
+	}
+
+	p.logger.Info().Msg("download step completed")
+
+	if p.canStartIndexing() {
+		p.transitionTo(StateIndexing)
+		return true
+	}
+
+	return true
```

Comment:
if downloading succeeded, but `canStartIndexing` returns false, this will not transition and then rerun the downloading logic. I think in the case when we can't transition to indexing, it should transition to canceled